bin/crew

```diff
@@ -556,77 +561,114 @@ def download
-
-    abort 'Checksum mismatch. :/ Try again.'.lightred unless
-      Digest::SHA256.hexdigest( File.read(filename) ) == sha256sum
-    puts "Archive downloaded".lightgreen
+      # Download file if not cached
+      system "#{CURL} --retry 3 -#{@verbose}#LC - --insecure \'#{url}\' --output #{filename}"
+      abort 'Checksum mismatch. :/ Try again.'.lightred unless
+        Digest::SHA256.hexdigest( File.read(filename) ) == sha256sum
+      puts "Archive downloaded".lightgreen
+      # Stow file in cache if requested
+      if CREW_IS_CACHING
+        puts 'Caching downloaded archive'.lightgreen
+        FileUtils.cp filename, CREW_CACHE_DIR, verbose: @fileutils_verbose
+      end
 
-    if ENV['CREW_CACHE_OPT']
-      puts 'Caching downloaded archive'.lightgreen
-      FileUtils.cp filename, CREW_CACHE_DIR, verbose: @fileutils_verbose
+    # Sources that download with git
+    when /\.git$/i
+      # Recall repository from cache if requested
+      if CREW_IS_CACHING
+        cachefile = CREW_CACHE_DIR + '/' + filename + '.tar.xz'
+        if File.file?(cachefile)
+          if system "sha256sum -c #{cachefile}.sha256"
+            FileUtils.mkdir $extract_dir
+            system "tar x#{@verbose}f #{cachefile} -C #{$extract_dir}"
+            return {source: source, filename: filename}
+          else
+            puts 'Cached repository checksum mismatch. :/ Will download.'.lightred
+          end
+        else
+          puts 'Cannot find cached repository. :/ Will download.'.lightred
+        end
+      end
+      # Download file if not needed
+      Dir.mkdir $extract_dir
+      Dir.chdir $extract_dir do
+        system 'git init'
+        system 'git config advice.detachedHead false'
+        system 'git config init.defaultBranch master'
+        system "git remote add origin #{@pkg.source_url}", exception: true
+        if @pkg.git_hash
+          system "git fetch origin"
+          system "git checkout #{@pkg.git_hash}", exception: true
+        else
+          system "git fetch --depth 1 origin #{@pkg.git_tag}", exception: true
+        end
+        system 'git checkout FETCH_HEAD'
+        puts "Repository downloaded".lightgreen
+      end
+      # Stow file in cache if requested
+      if CREW_IS_CACHING
+        puts 'Caching downloaded archive.'.lightgreen
+          Dir.chdir "#{$extract_dir}" do system "tar c#{@verbose}Jf #{CREW_CACHE_DIR}/#{filename}.tar.xz $(find -mindepth 1 -maxdepth 1 -printf '%P\n')" end
+        system "sha256sum #{CREW_CACHE_DIR}/#{filename}.tar.xz > #{CREW_CACHE_DIR}/#{filename}.tar.xz.sha256"
+      end
     end
   end
   return {source: source, filename: filename}
 end
 
 def unpack (meta)
-  extract_dir = meta[:filename] + '.' + SecureRandom.alphanumeric(8) + '.dir'
   target_dir = nil
   Dir.chdir CREW_BREW_DIR do
-    FileUtils.mkdir_p extract_dir, verbose: @fileutils_verbose
+    FileUtils.mkdir_p $extract_dir, verbose: @fileutils_verbose
```

Comment:
Why the `$`?  This is ruby, not bash.

Reply:
the $ makes the variable global.

Reply:
Why does it need to be global?

Reply:
Because it's called in `def download` and `def unpack`, and it needs to be the same since the variable contains entropy.

Reply:
Better to use instance variable (make it consistent with other variables)? It also shared between functions

Reply:
@supechicken, the problem is that the download function isn't in a class.

Reply:
@saltedcoffii: crew has worked for years without global variables.  Why do we suddenly need them now?

Reply:
It also works without class (see `set_package` section)

Reply:
Please look into fixing https://github.com/skycocker/chromebrew/issues/5603.  That should be a priority.

Reply:
Okay, I switched the $ to an @ and it works, thanks for the suggestion!